File: hmi/shared/contracts.ts
export type RuntimeMode = 'sim' | 'hardware' | 'unknown';
export type TransportState = 'connected' | 'connecting' | 'disconnected';
export type ConnectionHealth = 'healthy' | 'degraded' | 'down';
export type TelemetryState = 'fresh' | 'stale' | 'unavailable';
export type LeaseRole = 'controller' | 'observer';
export type CommandRiskLevel = 'low' | 'medium' | 'high' | 'critical';

export type CommandLifecycleState =
  | 'RECEIVED'
  | 'PARSING'
  | 'VALIDATING'
  | 'NEEDS_CONFIRMATION'
  | 'CONFIRMED'
  | 'EXECUTION_REQUESTED'
  | 'EXECUTING'
  | 'SUCCEEDED'
  | 'FAILED'
  | 'REJECTED'
  | 'CANCELLED'
  | 'EXPIRED';

export type TerminalCommandLifecycleState = 'SUCCEEDED' | 'FAILED' | 'REJECTED' | 'CANCELLED' | 'EXPIRED';

export type SystemRuntimeState =
  | 'NORMAL'
  | 'FAULT'
  | 'ESTOP'
  | 'HOLD'
  | 'TIMEOUT'
  | 'LOST_CONN'
  | 'SAFETY_BLOCKED';

export type MessageOrigin = 'system' | 'operator' | 'assistant';

export interface BridgeConnection {
  name: 'ros2' | 'moveit2' | 'llm' | 'motoros2';
  label: string;
  health: ConnectionHealth;
}

export interface LeaseView {
  leaseId: string | null;
  leaseToken: string | null;
  role: LeaseRole;
  ownsControl: boolean;
  holderOperatorId: string | null;
  holderSessionId: string | null;
  acquiredAt: string | null;
  expiresAt: string | null;
  statusText: string;
  canForceTakeover: boolean;
}

export interface BridgeCapabilities {
  readOnly: boolean;
  canAcquireLease: boolean;
  canSubmitCommands: boolean;
  canConfirmCommands: boolean;
  canCancelCommands: boolean;
  canAbortCommands: boolean;
  commandIngressAvailable: boolean;
  confirmationAvailable: boolean;
  executionAllowed: boolean;
  replayAvailable: boolean;
  simOnly: boolean;
  hardwareGate: HardwareGateStatus;
}

export interface HardwareGateChecklist {
  timingJitter: boolean;
  disconnectReconnect: boolean;
  robotStatusSemantics: boolean;
  jointSourcePrecedence: boolean;
  auditVisibility: boolean;
}

export interface HardwareGateStatus {
  unlocked: boolean;
  reasons: string[];
  flagEnabled: boolean;
  evidencePath: string;
  approvedBy: string | null;
  approvedAt: string | null;
  reportPath: string | null;
  reportSha256: string | null;
  reportSha256Match: boolean;
  checklist: HardwareGateChecklist | null;
}

export interface TelemetrySourceStatus {
  name: string;
  label: string;
  topic: string;
  lastSeenAt: string | null;
  freshnessThresholdSec: number;
  freshnessState: TelemetryState;
  preferred: boolean;
  active: boolean;
  detail: string | null;
}

export interface ValidationSourceStatus {
  name: string;
  label: string;
  topic: string;
  freshnessState: TelemetryState;
  active: boolean;
  preferred: boolean;
  detail: string | null;
}

export interface JointPosition {
  name: string;
  positionDeg: number | null;
  minDeg: number;
  maxDeg: number;
}

export interface PlanMetrics {
  score: number | null;
  pathLengthRad: number | null;
  smoothness: number | null;
  clearanceM: number | null;
  cartesianCompletionPct: number | null;
  replanCount: number | null;
}

export interface RobotStatusSnapshot {
  servoState: 'ON' | 'OFF' | 'UNKNOWN';
  eStop: 'CLEAR' | 'ACTIVE' | 'UNKNOWN';
  alarmState: 'NONE' | 'ACTIVE' | 'UNKNOWN';
  motionMode: string | null;
  trajectoryPointsUsed: number | null;
  trajectoryPointsCapacity: number | null;
  readinessMessage: string;
}

export interface RuntimeSnapshot {
  systemState: SystemRuntimeState;
  blocking: boolean;
  statusText: string;
  mode: RuntimeMode;
  robotStatus: RobotStatusSnapshot;
}

export interface ChatMessage {
  id: string;
  commandId: string | null;
  origin: MessageOrigin;
  timestamp: string;
  text: string;
  tag?: string | null;
}

export interface CommandValidationResult {
  accepted: boolean;
  leaseValid: boolean;
  runtimeAllowed: boolean;
  telemetryFresh: boolean;
  requiresConfirmation: boolean;
  riskLevel: CommandRiskLevel | null;
  blockingReasons: string[];
  confirmationReasons: string[];
  planFingerprint: string | null;
  executionAllowedNow: boolean;
  criticalSources: ValidationSourceStatus[];
  optionalSources: ValidationSourceStatus[];
  eventDrivenSources: ValidationSourceStatus[];
  hardwareGate: HardwareGateStatus;
  preflight: Record<string, unknown>;
}

export interface CommandExecutionResult {
  accepted: boolean;
  adapter: string;
  status: string;
  summary: string;
  dispatchedToRos: boolean;
  queryOnly?: boolean;
  referenceFrame?: string | null;
  pose?: Record<string, unknown> | null;
  poseMm?: Record<string, unknown> | null;
  commandId?: string | null;
  planFingerprint?: string | null;
  operatorId?: string | null;
  sessionId?: string | null;
  leaseId?: string | null;
  correlationId?: string | null;
}

export interface CommandView {
  commandId: string;
  commandKind: 'command';
  sessionId: string;
  operatorId: string;
  rawText: string;
  intentSource: 'text' | 'structured';
  structuredIntent?: Record<string, unknown> | null;
  lifecycleState: CommandLifecycleState;
  summaryLabel: string;
  plannerUsed: string | null;
  frameUsed: string | null;
  mode: RuntimeMode;
  riskLevel: CommandRiskLevel | null;
  planFingerprint: string | null;
  correlationId: string | null;
  rejectReason: string | null;
  parsedIntent?: Record<string, unknown> | null;
  validationResult?: CommandValidationResult | null;
  planSummary?: Record<string, unknown> | null;
  metrics?: PlanMetrics | null;
  confirmationExpiresAt: string | null;
  createdAt: string;
  confirmAt: string | null;
  executeAt: string | null;
  executionResult?: CommandExecutionResult | null;
  finalState: TerminalCommandLifecycleState | null;
  parentSequenceId?: string | null;
  sequenceStepIndex?: number | null;
  sequenceStepCount?: number | null;
}

export interface SequenceView {
  sequenceId: string;
  commandKind: 'sequence';
  sessionId: string;
  operatorId: string;
  rawText: string;
  intentSource: 'text' | 'structured';
  structuredIntent?: Record<string, unknown> | null;
  lifecycleState: CommandLifecycleState;
  summaryLabel: string;
  plannerUsed: string | null;
  frameUsed: string | null;
  mode: RuntimeMode;
  riskLevel: CommandRiskLevel | null;
  planFingerprint: string | null;
  correlationId: string | null;
  rejectReason: string | null;
  validationResult?: CommandValidationResult | null;
  planSummary?: Record<string, unknown> | null;
  metrics?: PlanMetrics | null;
  confirmationExpiresAt: string | null;
  createdAt: string;
  confirmAt: string | null;
  executeAt: string | null;
  executionResult?: CommandExecutionResult | null;
  finalState: TerminalCommandLifecycleState | null;
  stepCount: number;
  currentStepIndex?: number | null;
  diagnostics: string[];
  manualRecoveryRequired: boolean;
  steps: CommandView[];
}

export interface ReplayListItem {
  commandId: string;
  kind: 'command' | 'sequence';
  sessionId: string;
  operatorId: string;
  summaryLabel: string;
  lifecycleState: string;
  finalState: string | null;
  plannerUsed: string | null;
  frameUsed: string | null;
  mode: RuntimeMode;
  createdAt: string;
  executeAt: string | null;
  riskLevel: CommandRiskLevel | null;
  stepCount?: number | null;
  currentStepIndex?: number | null;
  manualRecoveryRequired: boolean;
}

export interface TimelineEvent {
  id: string;
  commandId: string | null;
  timestamp: string;
  fromState: string | null;
  toState: string | null;
  runtimeState: string | null;
  message: string;
  payload?: Record<string, unknown> | null;
}

export interface ReplayDetail {
  jobType: 'command' | 'sequence';
  command: CommandView | null;
  sequence: SequenceView | null;
  timeline: TimelineEvent[];
  runtimeEvents: TimelineEvent[];
}

export interface HmiStateSnapshot {
  schemaVersion: string;
  generatedAt: string;
  transportState: TransportState;
  telemetryState: TelemetryState;
  telemetrySources: TelemetrySourceStatus[];
  mode: RuntimeMode;
  connections: BridgeConnection[];
  capabilities: BridgeCapabilities;
  lease: LeaseView;
  runtime: RuntimeSnapshot;
  messages: ChatMessage[];
  activeCommand: CommandView | null;
  activeSequence: SequenceView | null;
  jointPositions: JointPosition[];
  planMetrics: PlanMetrics | null;
  replayItems: ReplayListItem[];
}

export interface RuntimeStateResponse {
  schemaVersion: string;
  generatedAt: string;
  telemetryState: TelemetryState;
  telemetrySources: TelemetrySourceStatus[];
  runtime: RuntimeSnapshot;
  jointPositions: JointPosition[];
}

export interface ConnectionStateResponse {
  schemaVersion: string;
  generatedAt: string;
  transportState: TransportState;
  telemetryState: TelemetryState;
  telemetrySources: TelemetrySourceStatus[];
  connections: BridgeConnection[];
}

export interface LeaseStateResponse {
  schemaVersion: string;
  generatedAt: string;
  capabilities: BridgeCapabilities;
  lease: LeaseView;
}

export interface LeaseAcquireRequest {
  sessionId: string;
  operatorId: string;
  requestedRole: LeaseRole;
  forceTakeover?: boolean;
  takeoverReason?: string;
}

export interface LeaseRenewRequest {
  sessionId: string;
  operatorId: string;
  leaseToken: string;
}

export interface LeaseReleaseRequest {
  sessionId: string;
  operatorId: string;
  leaseToken: string;
}

export interface CommandIntentRequest {
  sessionId: string;
  operatorId: string;
  leaseToken: string | null;
  intentText?: string | null;
  mode: 'sim' | 'hardware';
}

export interface CommandConfirmRequest {
  sessionId: string;
  operatorId: string;
  leaseToken: string | null;
  planFingerprint: string;
}

export interface CommandCancelRequest {
  sessionId: string;
  operatorId: string;
  leaseToken: string | null;
  reason?: string;
}

export interface LeaseMutationResponse {
  accepted: boolean;
  lease: LeaseView;
  reason: string | null;
}

export interface CommandMutationResponse {
  accepted: boolean;
  jobType: 'command' | 'sequence';
  commandId: string | null;
  sequenceId?: string | null;
  reason: string | null;
  snapshot?: HmiStateSnapshot | null;
  command: CommandView | null;
  sequence?: SequenceView | null;
}

export interface ReplayListQuery {
  sessionId?: string;
  operatorId?: string;
  finalState?: string;
  from?: string;
  to?: string;
  limit?: number;
}

export interface ReplayListResponse {
  items: ReplayListItem[];
}

export type HmiStreamEvent =
  | { type: 'snapshot'; snapshot: HmiStateSnapshot }
  | { type: 'heartbeat'; schemaVersion: string; generatedAt: string; transportState: TransportState; telemetryState: TelemetryState }
  | { type: 'lease_state'; lease: LeaseView; capabilities: BridgeCapabilities }
  | { type: 'command_lifecycle'; command: CommandView; messages?: ChatMessage[]; planMetrics?: PlanMetrics | null }
  | { type: 'sequence_lifecycle'; sequence: SequenceView; messages?: ChatMessage[] }
  | { type: 'replay_updated'; replayItems: ReplayListItem[] }
  | { type: 'connection_state'; transportState: TransportState; connections?: BridgeConnection[] }
  | { type: 'jog_bridge_status'; jogBridgeStatus: JogBridgeStatusSnapshot };

// ── Jog Pendant Types ──────────────────────────────────────────────────────

export type JogMode = 'continuous' | 'discrete';

export type JogBridgeState =
  | 'IDLE'
  | 'STARTING'
  | 'READY'
  | 'ACTIVE'
  | 'HALTING'
  | 'HALTED'
  | 'ERROR'
  | 'REJECTED_NOT_READY'
  | 'REJECTED_FJT_ACTIVE'
  | 'TIMEOUT'
  | 'BUSY_RETRY';

export interface JogBridgeStatusSnapshot {
  state: JogBridgeState;
  pointsQueued: number;
  effectiveHz: number;
  robotReady: boolean;
  servoActive: boolean;
  bridgeActive: boolean;
  lastError: string;
  rejectionReason: string;
}

export interface JogCommandRequest {
  jointIndex: number;       // 0-5
  direction: 1 | -1;        // +1 = positive, -1 = negative
  mode: JogMode;
  velocityScale: number;     // 0.0-0.3
  stepDegrees: number;       // for discrete mode
}

export interface JogBridgeCapabilities {
  jogAvailable: boolean;
  bridgeServiceAvailable: boolean;
  canActivateBridge: boolean;
  bridgeState: JogBridgeState;
  isExclusiveMode: boolean;  // true when jog bridge is active (blocks FJT path)
}

export interface ServoControlResponse {
  accepted: boolean;
  message: string;
}

export interface ServoControlRequest {
  sessionId: string;
  operatorId: string;
  leaseToken: string | null;
}

export interface GP4BridgeClient {
  connect(params: {
    sessionId: string;
    operatorId: string;
    onEvent: (event: HmiStreamEvent) => void;
    onTransportStateChange?: (state: TransportState) => void;
  }): () => void;
  acquireLease(request: LeaseAcquireRequest): Promise<LeaseMutationResponse>;
  renewLease(request: LeaseRenewRequest): Promise<LeaseMutationResponse>;
  releaseLease(request: LeaseReleaseRequest): Promise<LeaseMutationResponse>;
  submitCommand(request: CommandIntentRequest): Promise<CommandMutationResponse>;
  confirmCommand(commandId: string, request: CommandConfirmRequest): Promise<CommandMutationResponse>;
  confirmSequence(sequenceId: string, request: CommandConfirmRequest): Promise<CommandMutationResponse>;
  abortCommand(commandId: string, request: CommandCancelRequest): Promise<CommandMutationResponse>;
  abortSequence(sequenceId: string, request: CommandCancelRequest): Promise<CommandMutationResponse>;
  getRuntimeState(sessionId: string, operatorId: string): Promise<RuntimeStateResponse>;
  getConnectionState(): Promise<ConnectionStateResponse>;
  getLeaseState(sessionId: string, operatorId: string): Promise<LeaseStateResponse>;
  listReplay(query?: ReplayListQuery): Promise<ReplayListResponse>;
  getReplayDetail(commandId: string): Promise<ReplayDetail>;
  getSequence(sequenceId: string): Promise<SequenceView>;
  // Jog pendant
  activateJogBridge(): Promise<{ accepted: boolean; message: string }>;
  deactivateJogBridge(): Promise<{ accepted: boolean; message: string }>;
  sendJogCommand(cmd: JogCommandRequest): Promise<{ accepted: boolean; message: string }>;
  // Servo control
  startServo(request: ServoControlRequest): Promise<ServoControlResponse>;
  stopServo(request: ServoControlRequest): Promise<ServoControlResponse>;
}

File: hmi/backend/api/contracts.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, TypeAdapter, model_validator

from ..services.telemetry_bridge_service import SNAPSHOT_SCHEMA_VERSION


class StrictModel(BaseModel):
    model_config = ConfigDict(extra='forbid')


class BridgeConnectionModel(StrictModel):
    name: Literal['ros2', 'moveit2', 'llm', 'motoros2']
    label: str
    health: Literal['healthy', 'degraded', 'down']


class LeaseViewModel(StrictModel):
    leaseId: str | None
    leaseToken: str | None
    role: Literal['controller', 'observer']
    ownsControl: bool
    holderOperatorId: str | None
    holderSessionId: str | None
    acquiredAt: str | None
    expiresAt: str | None
    statusText: str
    canForceTakeover: bool


class BridgeCapabilitiesModel(StrictModel):
    readOnly: bool
    canAcquireLease: bool
    canSubmitCommands: bool
    canConfirmCommands: bool
    canCancelCommands: bool = False
    canAbortCommands: bool
    commandIngressAvailable: bool = False
    confirmationAvailable: bool = False
    executionAllowed: bool = False
    replayAvailable: bool = False
    simOnly: bool = False
    hardwareGate: HardwareGateStatusModel


class HardwareGateChecklistModel(StrictModel):
    timingJitter: bool
    disconnectReconnect: bool
    robotStatusSemantics: bool
    jointSourcePrecedence: bool
    auditVisibility: bool


class HardwareGateStatusModel(StrictModel):
    unlocked: bool
    reasons: list[str]
    flagEnabled: bool
    evidencePath: str
    approvedBy: str | None
    approvedAt: str | None
    reportPath: str | None
    reportSha256: str | None
    reportSha256Match: bool
    checklist: HardwareGateChecklistModel | None


class JointPositionModel(StrictModel):
    name: str
    positionDeg: float | None
    minDeg: float
    maxDeg: float


class TelemetrySourceStatusModel(StrictModel):
    name: str
    label: str
    topic: str
    lastSeenAt: str | None
    freshnessThresholdSec: float
    freshnessState: Literal['fresh', 'stale', 'unavailable']
    preferred: bool
    active: bool
    detail: str | None


class RobotStatusSnapshotModel(StrictModel):
    servoState: Literal['ON', 'OFF', 'UNKNOWN']
    eStop: Literal['CLEAR', 'ACTIVE', 'UNKNOWN']
    alarmState: Literal['NONE', 'ACTIVE', 'UNKNOWN']
    motionMode: str | None
    trajectoryPointsUsed: int | None
    trajectoryPointsCapacity: int | None
    readinessMessage: str


class RuntimeSnapshotModel(StrictModel):
    systemState: Literal['NORMAL', 'FAULT', 'ESTOP', 'HOLD', 'TIMEOUT', 'LOST_CONN', 'SAFETY_BLOCKED']
    blocking: bool
    statusText: str
    mode: Literal['sim', 'hardware', 'unknown']
    robotStatus: RobotStatusSnapshotModel


class ChatMessageModel(StrictModel):
    id: str
    commandId: str | None
    origin: Literal['system', 'operator', 'assistant']
    timestamp: str
    text: str
    tag: str | None = None


class PlanMetricsModel(StrictModel):
    score: float | None
    pathLengthRad: float | None
    smoothness: float | None
    clearanceM: float | None
    cartesianCompletionPct: float | None
    replanCount: int | None


class ValidationSourceStatusModel(StrictModel):
    name: str
    label: str
    topic: str
    freshnessState: Literal['fresh', 'stale', 'unavailable']
    active: bool
    preferred: bool
    detail: str | None


class CommandValidationResultModel(StrictModel):
    accepted: bool
    leaseValid: bool
    runtimeAllowed: bool
    telemetryFresh: bool
    requiresConfirmation: bool
    riskLevel: Literal['low', 'medium', 'high', 'critical'] | None
    blockingReasons: list[str]
    confirmationReasons: list[str]
    planFingerprint: str | None
    executionAllowedNow: bool
    criticalSources: list[ValidationSourceStatusModel]
    optionalSources: list[ValidationSourceStatusModel]
    eventDrivenSources: list[ValidationSourceStatusModel]
    hardwareGate: HardwareGateStatusModel
    preflight: dict[str, Any]


class CommandExecutionResultModel(StrictModel):
    accepted: bool
    adapter: str
    status: str
    summary: str
    dispatchedToRos: bool
    commandId: str | None = None
    planFingerprint: str | None = None
    operatorId: str | None = None
    sessionId: str | None = None
    leaseId: str | None = None
    correlationId: str | None = None


class CommandViewModel(StrictModel):
    commandId: str
    commandKind: Literal['command']
    sessionId: str
    operatorId: str
    rawText: str
    intentSource: Literal['text', 'structured']
    structuredIntent: dict[str, Any] | None = None
    lifecycleState: Literal[
        'RECEIVED',
        'PARSING',
        'VALIDATING',
        'NEEDS_CONFIRMATION',
        'CONFIRMED',
        'EXECUTION_REQUESTED',
        'EXECUTING',
        'SUCCEEDED',
        'FAILED',
        'REJECTED',
        'CANCELLED',
        'EXPIRED',
    ]
    summaryLabel: str
    plannerUsed: str | None
    frameUsed: str | None
    mode: Literal['sim', 'hardware', 'unknown']
    riskLevel: Literal['low', 'medium', 'high', 'critical'] | None = None
    planFingerprint: str | None = None
    correlationId: str | None = None
    rejectReason: str | None
    parsedIntent: dict[str, Any] | None = None
    validationResult: CommandValidationResultModel | None = None
    planSummary: dict[str, Any] | None = None
    metrics: PlanMetricsModel | None = None
    confirmationExpiresAt: str | None = None
    createdAt: str
    confirmAt: str | None
    executeAt: str | None
    executionResult: CommandExecutionResultModel | None = None
    finalState: Literal['SUCCEEDED', 'FAILED', 'REJECTED', 'CANCELLED', 'EXPIRED'] | None = None
    parentSequenceId: str | None = None
    sequenceStepIndex: int | None = None
    sequenceStepCount: int | None = None


class SequenceViewModel(StrictModel):
    sequenceId: str
    commandKind: Literal['sequence']
    sessionId: str
    operatorId: str
    rawText: str
    intentSource: Literal['text', 'structured']
    structuredIntent: dict[str, Any] | None = None
    lifecycleState: Literal[
        'RECEIVED',
        'PARSING',
        'VALIDATING',
        'NEEDS_CONFIRMATION',
        'CONFIRMED',
        'EXECUTION_REQUESTED',
        'EXECUTING',
        'SUCCEEDED',
        'FAILED',
        'REJECTED',
        'CANCELLED',
        'EXPIRED',
    ]
    summaryLabel: str
    plannerUsed: str | None
    frameUsed: str | None
    mode: Literal['sim', 'hardware', 'unknown']
    riskLevel: Literal['low', 'medium', 'high', 'critical'] | None = None
    planFingerprint: str | None = None
    correlationId: str | None = None
    rejectReason: str | None
    validationResult: CommandValidationResultModel | None = None
    planSummary: dict[str, Any] | None = None
    metrics: PlanMetricsModel | None = None
    confirmationExpiresAt: str | None = None
    createdAt: str
    confirmAt: str | None
    executeAt: str | None
    executionResult: CommandExecutionResultModel | None = None
    finalState: Literal['SUCCEEDED', 'FAILED', 'REJECTED', 'CANCELLED', 'EXPIRED'] | None = None
    stepCount: int
    currentStepIndex: int | None = None
    diagnostics: list[str] = Field(default_factory=list)
    manualRecoveryRequired: bool = False
    steps: list[CommandViewModel] = Field(default_factory=list)


class ReplayListItemModel(StrictModel):
    commandId: str
    kind: Literal['command', 'sequence']
    sessionId: str
    operatorId: str
    summaryLabel: str
    lifecycleState: str
    finalState: str | None
    plannerUsed: str | None
    frameUsed: str | None
    mode: Literal['sim', 'hardware', 'unknown']
    createdAt: str
    executeAt: str | None
    riskLevel: Literal['low', 'medium', 'high', 'critical'] | None = None
    stepCount: int | None = None
    currentStepIndex: int | None = None
    manualRecoveryRequired: bool = False


class TimelineEventModel(StrictModel):
    id: str
    commandId: str | None
    timestamp: str
    fromState: str | None
    toState: str | None
    runtimeState: str | None
    message: str
    payload: dict[str, Any] | None = None


class ReplayDetailModel(StrictModel):
    jobType: Literal['command', 'sequence']
    command: CommandViewModel | None = None
    sequence: SequenceViewModel | None = None
    timeline: list[TimelineEventModel]
    runtimeEvents: list[TimelineEventModel]


class HmiStateSnapshotModel(StrictModel):
    schemaVersion: Literal[SNAPSHOT_SCHEMA_VERSION]
    generatedAt: str
    transportState: Literal['connected', 'connecting', 'disconnected']
    telemetryState: Literal['fresh', 'stale', 'unavailable']
    telemetrySources: list[TelemetrySourceStatusModel]
    mode: Literal['sim', 'hardware', 'unknown']
    connections: list[BridgeConnectionModel]
    capabilities: BridgeCapabilitiesModel
    lease: LeaseViewModel
    runtime: RuntimeSnapshotModel
    messages: list[ChatMessageModel]
    activeCommand: CommandViewModel | None
    activeSequence: SequenceViewModel | None = None
    jointPositions: list[JointPositionModel]
    planMetrics: PlanMetricsModel | None
    replayItems: list[ReplayListItemModel]


class RuntimeStateResponseModel(StrictModel):
    schemaVersion: Literal[SNAPSHOT_SCHEMA_VERSION]
    generatedAt: str
    telemetryState: Literal['fresh', 'stale', 'unavailable']
    telemetrySources: list[TelemetrySourceStatusModel]
    runtime: RuntimeSnapshotModel
    jointPositions: list[JointPositionModel]


class ConnectionStateResponseModel(StrictModel):
    schemaVersion: Literal[SNAPSHOT_SCHEMA_VERSION]
    generatedAt: str
    transportState: Literal['connected', 'connecting', 'disconnected']
    telemetryState: Literal['fresh', 'stale', 'unavailable']
    telemetrySources: list[TelemetrySourceStatusModel]
    connections: list[BridgeConnectionModel]


class LeaseStateResponseModel(StrictModel):
    schemaVersion: Literal[SNAPSHOT_SCHEMA_VERSION]
    generatedAt: str
    capabilities: BridgeCapabilitiesModel
    lease: LeaseViewModel


class LeaseAcquireRequestModel(StrictModel):
    sessionId: str
    operatorId: str
    requestedRole: Literal['controller', 'observer']
    forceTakeover: bool = False
    takeoverReason: str | None = None


class LeaseRenewRequestModel(StrictModel):
    sessionId: str
    operatorId: str
    leaseToken: str


class LeaseReleaseRequestModel(StrictModel):
    sessionId: str
    operatorId: str
    leaseToken: str


class LeaseMutationResponseModel(StrictModel):
    accepted: bool
    lease: LeaseViewModel
    reason: str | None


class CommandIntentRequestModel(StrictModel):
    sessionId: str
    operatorId: str
    leaseToken: str | None
    intentText: str | None = None
    structuredIntent: dict[str, Any] | None = None
    mode: Literal['sim', 'hardware', 'unknown']

    @model_validator(mode='after')
    def validate_intent_payload(self) -> 'CommandIntentRequestModel':
        if not (self.intentText and self.intentText.strip()) and self.structuredIntent is None:
            raise ValueError('intentText or structuredIntent is required')
        return self


class CommandConfirmRequestModel(StrictModel):
    sessionId: str
    operatorId: str
    leaseToken: str | None
    planFingerprint: str


class CommandCancelRequestModel(StrictModel):
    sessionId: str
    operatorId: str
    leaseToken: str | None
    reason: str | None = None


class CommandMutationResponseModel(StrictModel):
    accepted: bool
    jobType: Literal['command', 'sequence']
    commandId: str | None
    sequenceId: str | None = None
    reason: str | None
    snapshot: HmiStateSnapshotModel | None = None
    command: CommandViewModel | None = None
    sequence: SequenceViewModel | None = None


class CommandListResponseModel(StrictModel):
    items: list[ReplayListItemModel]


class SnapshotStreamEventModel(StrictModel):
    type: Literal['snapshot']
    snapshot: HmiStateSnapshotModel


class HeartbeatStreamEventModel(StrictModel):
    type: Literal['heartbeat']
    schemaVersion: Literal[SNAPSHOT_SCHEMA_VERSION]
    generatedAt: str
    transportState: Literal['connected', 'connecting', 'disconnected']
    telemetryState: Literal['fresh', 'stale', 'unavailable']


class LeaseStateStreamEventModel(StrictModel):
    type: Literal['lease_state']
    lease: LeaseViewModel
    capabilities: BridgeCapabilitiesModel


class CommandLifecycleStreamEventModel(StrictModel):
    type: Literal['command_lifecycle']
    command: CommandViewModel
    messages: list[ChatMessageModel] = Field(default_factory=list)
    planMetrics: PlanMetricsModel | None = None


class SequenceLifecycleStreamEventModel(StrictModel):
    type: Literal['sequence_lifecycle']
    sequence: SequenceViewModel
    messages: list[ChatMessageModel] = Field(default_factory=list)


class ReplayUpdatedStreamEventModel(StrictModel):
    type: Literal['replay_updated']
    replayItems: list[ReplayListItemModel]


# ── Jog Pendant Models ──────────────────────────────────────────────────────

class JogBridgeStatusModel(StrictModel):
    state: str
    pointsQueued: int
    effectiveHz: float
    robotReady: bool
    servoActive: bool
    bridgeActive: bool
    lastError: str
    rejectionReason: str


class JogBridgeStatusStreamEventModel(StrictModel):
    type: Literal['jog_bridge_status']
    jogBridgeStatus: JogBridgeStatusModel


class JogCommandRequestModel(StrictModel):
    jointIndex: int = Field(..., ge=0, le=5)
    direction: Literal[-1, 1]
    mode: Literal['continuous', 'discrete']
    velocityScale: float = Field(..., ge=0.0, le=0.3)
    stepDegrees: float = Field(0.0, ge=0.0, le=10.0)


HMI_STREAM_EVENT_ADAPTER = TypeAdapter(
    SnapshotStreamEventModel
    | HeartbeatStreamEventModel
    | LeaseStateStreamEventModel
    | CommandLifecycleStreamEventModel
    | SequenceLifecycleStreamEventModel
    | ReplayUpdatedStreamEventModel
    | JogBridgeStatusStreamEventModel
)
